src/net/mqtt_interface.rs

```diff
@@ -0,0 +1,235 @@
+use crate::hardware::{
+    design_parameters::MQTT_BROKER, CycleCounter, EthernetPhy, NetworkStack,
+};
+
+use core::{cell::RefCell, fmt::Write};
+
+use heapless::{consts, String};
+use serde::Serialize;
+
+use super::{Action, MqttMessage, SettingsResponse, SettingsResponseCode};
+
+/// MQTT settings interface.
+pub struct MqttInterface<S>
+where
+    S: miniconf::Miniconf + Default + Clone,
+{
+    telemetry_topic: String<consts::U128>,
+    default_response_topic: String<consts::U128>,
+    mqtt: RefCell<minimq::MqttClient<minimq::consts::U256, NetworkStack>>,
+    settings: RefCell<S>,
+    clock: CycleCounter,
+    phy: EthernetPhy,
+    network_was_reset: bool,
+    subscribed: bool,
+    id: String<consts::U64>,
+}
+
+impl<S> MqttInterface<S>
+where
+    S: miniconf::Miniconf + Default + Clone,
+{
+    /// Construct a new MQTT settings interface.
+    ///
+    /// # Args
+    /// * `stack` - The network stack to use for communication.
+    /// * `client_id` - The ID of the MQTT client. May be an empty string for auto-assigning.
+    /// * `prefix` - The MQTT device prefix to use for this device.
+    /// * `phy` - The PHY driver for querying the link state.
+    /// * `clock` - The clock to utilize for querying the current system time.
+    pub fn new(
+        stack: NetworkStack,
+        client_id: &str,
+        prefix: &str,
+        phy: EthernetPhy,
+        clock: CycleCounter,
+    ) -> Self {
+        let mqtt_client =
+            minimq::MqttClient::new(MQTT_BROKER.into(), client_id, stack)
+                .unwrap();
+
+        let mut telemetry_topic: String<consts::U128> = String::new();
+        write!(&mut telemetry_topic, "{}/telemetry", prefix).unwrap();
+
+        let mut response_topic: String<consts::U128> = String::new();
+        write!(&mut response_topic, "{}/log", prefix).unwrap();
+
+        Self {
+            mqtt: RefCell::new(mqtt_client),
+            settings: RefCell::new(S::default()),
+            id: String::from(prefix),
```

Comment:
Why the conversion and not just `to_owned`? You need `str` afterwards anyway.

Reply:
heapless::String does not implement `ToOwned`